x/exchange/events_test.go

```diff
@@ -716,6 +729,522 @@ func TestNewEventParamsUpdated(t *testing.T) {
+func newTestPayment(t *testing.T, source, sourceAmount, target, targetAmount, externalID string) *Payment {
+	t.Helper()
+	rv := &Payment{
+		Source:     source,
+		Target:     target,
+		ExternalId: externalID,
+	}
+	var err error
+	rv.SourceAmount, err = sdk.ParseCoinsNormalized(sourceAmount)
+	require.NoError(t, err, "source amount: ParseCoinsNormalized(%q)", sourceAmount)
+	rv.TargetAmount, err = sdk.ParseCoinsNormalized(targetAmount)
+	require.NoError(t, err, "target amount: ParseCoinsNormalized(%q)", targetAmount)
+	return rv
+}
+
+func TestEventPaymentCreated(t *testing.T) {
+	tests := []struct {
+		name      string
+		payment   *Payment
+		expected  *EventPaymentCreated
+		expAllSet bool
+	}{
+		{
+			name:    "all payment fields have content",
+			payment: newTestPayment(t, "source_addr", "312strawberry", "target_addr", "7tangerine", "just_some_identifier"),
+			expected: &EventPaymentCreated{
+				Source:       "source_addr",
+				SourceAmount: "312strawberry",
+				Target:       "target_addr",
+				TargetAmount: "7tangerine",
+				ExternalId:   "just_some_identifier",
+			},
+			expAllSet: true,
+		},
+		{
+			name:    "zero source amount",
+			payment: newTestPayment(t, "source_addr", "", "target_addr", "7tangerine", "just_some_identifier"),
+			expected: &EventPaymentCreated{
+				Source:       "source_addr",
+				SourceAmount: "",
+				Target:       "target_addr",
+				TargetAmount: "7tangerine",
+				ExternalId:   "just_some_identifier",
+			},
+		},
+		{
+			name:    "no target",
+			payment: newTestPayment(t, "source_addr", "312strawberry", "", "7tangerine", "just_some_identifier"),
+			expected: &EventPaymentCreated{
+				Source:       "source_addr",
+				SourceAmount: "312strawberry",
+				Target:       "",
+				TargetAmount: "7tangerine",
+				ExternalId:   "just_some_identifier",
+			},
+		},
+		{
+			name:    "zero target amount",
+			payment: newTestPayment(t, "source_addr", "312strawberry", "target_addr", "", "just_some_identifier"),
+			expected: &EventPaymentCreated{
+				Source:       "source_addr",
+				SourceAmount: "312strawberry",
+				Target:       "target_addr",
+				TargetAmount: "",
+				ExternalId:   "just_some_identifier",
+			},
+		},
+		{
+			name:    "empty external id",
+			payment: newTestPayment(t, "source_addr", "312strawberry", "target_addr", "7tangerine", ""),
+			expected: &EventPaymentCreated{
+				Source:       "source_addr",
+				SourceAmount: "312strawberry",
+				Target:       "target_addr",
+				TargetAmount: "7tangerine",
+				ExternalId:   "",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var event *EventPaymentCreated
+			testFunc := func() {
+				event = NewEventPaymentCreated(tc.payment)
+			}
+			require.NotPanics(t, testFunc, "NewEventPaymentCreated")
+			assert.Equal(t, tc.expected, event, "NewEventPaymentCreated result")
+			assertEventContent(t, event, "EventPaymentCreated", tc.expAllSet)
+		})
+	}
+}
```

Comment:
The behavior of coins (invalid or otherwise) are well tested elsewhere. Since `NewEventPaymentCreated` doesn't do any special handling regarding invalid coins, this isn't the place to put unit tests involving invalid coins.

While it might be possible for `sdk.Coins.String` to panic, we just don't care about that behavior here. If it panics on invalid coins, `NewEventPaymentCreated` should panic too. If it gracefully handles invalid coins, that's fine too. But we don't care here what happens either way.

Futher, the SDK is making changes to `sdk.Coins` to hopefully make them less panic-happy. If I had a unit test here that asserted that `sdk.Coins.String` panics on invalid coins, that test would then fail once we update the SDK even though it's still behaving as expected. So it just doesn't makes sense to add such a test case.